addons/website_version/models/ir_ui_view.py

```diff
@@ -0,0 +1,146 @@
+
+    @tools.ormcache_context(**_read_template_cache)
+    def _read_template(self, cr, uid, view_id, context=None):
+        arch = self.read_combined(cr, uid, view_id, fields=['arch'], context=context)['arch']
+        arch_tree = etree.fromstring(arch)
+
+        if 'lang' in context:
+            arch_tree = self.translate_qweb(cr, uid, view_id, arch_tree, context['lang'], context)
+
+        self.distribute_branding(arch_tree)
+        root = etree.Element('templates')
+        root.append(arch_tree)
+        arch = etree.tostring(root, encoding='utf-8', xml_declaration=True)
+        return arch
+
+    @tools.ormcache(size=0)
+    def read_template(self, cr, uid, xml_id, context=None):
+        if isinstance(xml_id, (int, long)):
+            view_id = xml_id
+        else:
+            if '.' not in xml_id:
+                raise ValueError('Invalid template id: %r' % (xml_id,))
+            view_id = self.get_view_id(cr, uid, xml_id, context=context)
+        return self._read_template(cr, uid, view_id, context=context)
+
+    def clear_cache(self):
+        self._read_template.clear_cache(self)
+
+    #To take the right inheriting views
+    @api.model
+    def get_inheriting_views_arch(self, view_id, model):
+        arch = super(view, self).get_inheriting_views_arch(view_id, model)
+        vw = self.browse(view_id)
+        if not (self.env.context and self.env.context.get('website_id') and vw.type == 'qweb'):
+            return arch
+        #right_ids to collect the right ids according to the key
+        right_ids = {}
+        priority = {}
+        #To create a dico called view_arch where v(key) is the id and a the arch(value)
+        view_arch = dict([(v, a) for a, v in arch])
+        keys = self.browse(view_arch.keys()).read(['key', 'version_id', 'website_id'])
+        for k in keys:
+            #The view to take depends of the context
+            if self.env.context.get('version_id'):
+                #priority:1 take the view which is in the same version
+                if k['version_id'] and k['version_id'][0] == self.env.context.get('version_id'):
+                    right_ids[k['key']] = k['id']
+                    priority[k['key']] = 3
+                #priority:2 take the view which is just in the same website
+                elif k['version_id'] is False and k['website_id'] and k['website_id'][0] == self.env.context.get('website_id'):
+                    if not priority.get(k['key']) or priority.get(k['key']) < 3:
+                        right_ids[k['key']] = k['id']
+                        priority[k['key']] = 2
+                #priority:3 take the original view
+                elif k['version_id'] is False and k['website_id'] is False:
+                    if not priority.get(k['key']) or priority.get(k['key']) < 2:
+                        right_ids[k['key']] = k['id']
+                        priority[k['key']] = 1
+            else:
```

Comment:
Duplicated code with just some bits that change. I think you should handle directly context_version_id and not perform an if/else to simplify the code.